### PR TITLE
SRT

### PR DESCRIPTION
**Pull request by funman**
SRT protocol

### CODE REVIEW

**Comment by funman** [comment]
I hacked rist_ts and rist_rx changes to test quickly

Not sure if I should remove the changes completely or add new programs

**Comment by funman** [comment]
This depends on https://github.com/funman/bitstream/tree/srt

File: lib/upipe-srt/upipe_srt_handshake.c [hunk 1]
```diff
@@ -0,0 +1,1336 @@
...
+    upipe_srt_handshake_init_ubuf_mgr(upipe);
+    upipe_srt_handshake_init_output(upipe);
+
+    upipe_srt_handshake_init_upump_mgr(upipe);
+    upipe_srt_handshake_init_upump_timer(upipe);
+    upipe_srt_handshake_init_upump_timeout(upipe);
+    upipe_srt_handshake_init_uclock(upipe);
+    upipe_srt_handshake_require_uclock(upipe);
+
+    srand48((long)&uprobe); // FIXME
```
[the remark this answers is not in the document]
Comment (cmassiot) [reply lib/upipe-srt/upipe_srt_handshake.c:340]: It should probably be done by the upipe_mgr during the initalisation phase. The seed could be supplied by the application, so you don't have to bother with it.

**Review by kierank — COMMENTED** [review]
Needs to be keepalive AND ACKS for srt shutdown.

File: lib/upipe-srt/upipe_srt_receiver.c [hunk 2]
```diff
@@ -575,7 +576,7 @@ static void upipe_srt_receiver_timer(struct upump *upump)
         if (unlikely(!ubase_check(uref_clock_get_cr_sys(uref, &cr_sys))))
             upipe_warn_va(upipe, "Couldn't read cr_sys in %s()", __func__);
 
-        if (now - cr_sys <= upipe_srt_receiver->latency)
+        if (now - cr_sys <= upipe_srt_receiver->latency - rtt)
```
Comment (kierank) [inline lib/upipe-srt/upipe_srt_receiver.c:579]: Need to handle the case where rtt > latency otherwise right hand side will go to UINT64_MAX and packets will buffer infinitely.

**Review by kierank — COMMENTED** [review]
This patch breaks data transmission

**Comment by nto** [comment]
> I hacked rist_ts and rist_rx changes to test quickly
> 
> Not sure if I should remove the changes completely or add new programs

You should probably add `srt_tx` and `srt_rx`.

File: lib/upipe-modules/Makefile.am [hunk 3]
```diff
@@ -110,6 +110,11 @@ endif
 libupipe_modules_la_CPPFLAGS = -I$(top_builddir)/include -I$(top_srcdir)/include
 libupipe_modules_la_LIBADD = -lm $(top_builddir)/lib/upipe/libupipe.la
 
+if HAVE_GCRYPT
+libupipe_modules_la_CPPFLAGS += $(GCRYPT_CFLAGS)
+libupipe_modules_la_LIBADD += $(GCRYPT_LIBS)
+endif
+
```
Comment (nto) [inline lib/upipe-modules/Makefile.am:117]: This can be removed, as the implementation is no longer in `upipe-modules`.

File: lib/upipe-srt/Makefile.am [hunk 4]
```diff
@@ -0,0 +1,18 @@
+lib_LTLIBRARIES = libupipe_srt.la
+
+libupipe_srt_la_SOURCES = upipe_srt_handshake.c \
+	upipe_srt_sender.c \
+	upipe_srt_receiver.c
+
+libupipe_srt_la_CPPFLAGS = -I$(top_builddir) -I$(top_builddir)/include -I$(top_srcdir)/include
+libupipe_srt_la_CFLAGS = $(AM_CFLAGS) $(SRT_FLAGS) $(BITSTREAM_CFLAGS)
+libupipe_srt_la_LIBADD = $(top_builddir)/lib/upipe/libupipe.la $(SRT_LIBS)
```
Comment (nto) [inline lib/upipe-srt/Makefile.am:9]: `SRT_FLAGS` and `SRT_LIBS` refer to libsrt pkg-config variables.

File: include/Makefile.am [hunk 5]
```diff
@@ -105,5 +105,5 @@ SUBDIRS += upipe-bearssl
 endif
 
 if HAVE_SRT
-SUBDIRS += upump-srt
+SUBDIRS += upump-srt upipe-srt
```
Comment (nto) [inline include/Makefile.am:108]: `HAVE_SRT` is libsrt dependency. You can just enable upipe-srt unconditionally.

**Comment by funman** [comment]
> @funman [srth: no need to update flow def immediately](https://github.com/Upipe/upipe/pull/947/commits/15e3534afd7686d6573114f959e29d115fabead4) breaks data transmission

This seems to break rist_tx caller

File: examples/Makefile.am [hunk 6]
```diff
@@ -19,11 +19,17 @@ UPIPEOSX_LIBS = $(top_builddir)/lib/upipe-osx/libupipe_osx.la
...
 UPIPEDVB_LIBS = $(top_builddir)/lib/upipe-dvb/libupipe_dvb.la
 UPIPEBEARSSL_LIBS = $(top_builddir)/lib/upipe-bearssl/libupipe_bearssl.la
+UPIPESRT_LIBS = $(top_builddir)/lib/upipe-srt/libupipe_srt.la
 
 noinst_PROGRAMS = 
 
 fec_LDADD = $(LDADD) $(UPUMPEV_LIBS) $(UPIPEMODULES_LIBS) $(UPIPETS_LIBS)
-rist_rx_LDADD = $(LDADD) $(UPUMPEV_LIBS) $(UPIPEMODULES_LIBS) $(UPIPEFILTERS_LIBS)
+rist_rx_LDADD = $(LDADD) $(UPUMPEV_LIBS) $(UPIPEMODULES_LIBS) $(UPIPEFILTERS_LIBS) $(UPIPESRT_LIBS)
+rist_rx_CFLAGS = $(AM_CFLAGS) $(BITSTREAM_CFLAGS)
```
Comment (nto) [inline examples/Makefile.am:28]: These changes to `rist_rx_LDADD` and `rist_rx_CFLAGS` are no longer necessary.

**Comment by funman** [comment]
> There are still some indentation changes left in `rist_rx` and `rist_tx`.

Sorry I think I used git diff -w

File: examples/Makefile.am [hunk 7]
```diff
@@ -19,11 +19,16 @@ UPIPEOSX_LIBS = $(top_builddir)/lib/upipe-osx/libupipe_osx.la
...
+UPIPESRT_LIBS = $(top_builddir)/lib/upipe-srt/libupipe_srt.la
 
 noinst_PROGRAMS = 
 
 fec_LDADD = $(LDADD) $(UPUMPEV_LIBS) $(UPIPEMODULES_LIBS) $(UPIPETS_LIBS)
 rist_rx_LDADD = $(LDADD) $(UPUMPEV_LIBS) $(UPIPEMODULES_LIBS) $(UPIPEFILTERS_LIBS)
+srt_rx_LDADD = $(LDADD) $(UPUMPEV_LIBS) $(UPIPEMODULES_LIBS) $(UPIPEFILTERS_LIBS) $(UPIPESRT_LIBS)
+srt_rx_CFLAGS = $(AM_CFLAGS) $(BITSTREAM_CFLAGS)
+srt_tx_LDADD = $(LDADD) $(UPUMPEV_LIBS) $(UPIPEMODULES_LIBS) $(UPIPEFILTERS_LIBS) $(UPIPESRT_LIBS)
+srt_tx_CFLAGS = $(AM_CFLAGS) $(BITSTREAM_CFLAGS)
```
Comment (nto) [inline examples/Makefile.am:31]: I don't see any bitstream usage in `srt_rx` and `srt_tx`? Same for upipe-filters?

File: examples/srt_tx.c [hunk 8]
```diff
@@ -0,0 +1,438 @@
+/*
+ * Copyright (C) 2016-2017 Open Broadcast Systems Ltd.
```
Comment (nto) [inline examples/srt_tx.c:2]: Wrong copyright dates?

File: examples/srt_rx.c [hunk 9]
```diff
@@ -0,0 +1,480 @@
+/*
+ * Copyright (C) 2016-2017 Open Broadcast Systems Ltd.
```
Comment (nto) [inline examples/srt_rx.c:2]: Wrong copyright dates?

**Comment by nto** [comment]
What about adding `upipe_srt_source` and `upipe_srt_sink` pipes which could be used as replacement of `upipe_udp_source`/`upipe_rtp_source` and `upipe_udp_sink`? These would make `srt_tx`/`srt_rx` much simpler, and srt usage in client applications as well.

**Review by kierank — COMMENTED** [review]
The buffer exhaustion patch seems to cause variable latency issues in obd

**Comment by funman** [comment]
> What about adding `upipe_srt_source` and `upipe_srt_sink` pipes which could be used as replacement of `upipe_udp_source`/`upipe_rtp_source` and `upipe_udp_sink`? These would make `srt_tx`/`srt_rx` much simpler, and srt usage in client applications as well.

I don't understand what they would replace exactly, why does udpsink need to be replaced?

**Comment by nto** [comment]
> I don't understand what they would replace exactly, why does udpsink need to be replaced?

The same way you could replace `upipe_udp_source` with `upipe_rtp_source` to use RTP instead of UDP, you should be able to use `upipe_srt_source` (with parameters passed in the URI) to use SRT instead.

**Comment by funman** [comment]
> > I don't understand what they would replace exactly, why does udpsink need to be replaced?
> 
> The same way you could replace `upipe_udp_source` with `upipe_rtp_source` to use RTP instead of UDP, you should be able to use `upipe_srt_source` (with parameters passed in the URI) to use SRT instead.

I guess it would have to handle the creation of udpsink and srt handshake as well then?

**Comment by nto** [comment]
Yes, it encapsulates everything, which is why `srt_tx` and `srt_rx` will be much simpler afterwards.

**Comment by funman** [comment]
It's not simpler for me since they are done already :)

**Comment by funman** [comment]
> Yes, it encapsulates everything, which is why `srt_tx` and `srt_rx` will be much simpler afterwards.

Also wouldn't we be missing events between internal pipes that could be useful?

**Comment by kierank** [comment]
Maybe once we have fixed the bugs we could have a wrapper pipe?

**Review by kierank — COMMENTED** [review]
Nuke the rtt patches imo

**Comment by nto** [comment]
> Maybe once we have fixed the bugs we could have a wrapper pipe?

Agreed. It does not have to be in this pull request.

File: lib/upipe-srt/libupipe_srt.pc.in [hunk 10]
```diff
@@ -0,0 +1,11 @@
+prefix=@prefix@
+exec_prefix=@exec_prefix@
+libdir=@libdir@
+includedir=@includedir@
+Name: libupipe_srt
+Description: Upipe multimedia framework, libsrt interface module
```
Comment (nto) [inline lib/upipe-srt/libupipe_srt.pc.in:6]: Should be `SRT modules` instead of `libsrt interface module`.

**Review by kierank — COMMENTED** [review]
srts: delete old key from memory 

Not sure this is correct. The key might be needed for retransmissions. Need to wait the 0x4000 packets or whatever they specify.

**Comment by kierank** [comment]
@cmassiot @nto @quarium I think this is "done"

File: examples/rist_rx.c [hunk 11]
```diff
@@ -331,7 +331,7 @@ int main(int argc, char *argv[])
 
     udp_sink_mgr = upipe_udpsink_mgr_alloc();
 
-    uclock = uclock_std_alloc(UCLOCK_FLAG_REALTIME);
+    uclock = uclock_std_alloc(0);
```
Comment (JDarnley) [inline examples/rist_rx.c:334]: Did you intend this change and the similar one in rist_tx to remain in the PR?

**Comment by funman** [comment]
No, I think it should be on its own

File: examples/srt_rx.c [hunk 12]
```diff
@@ -0,0 +1,496 @@
...
+#define UDICT_POOL_DEPTH 10
+#define UREF_POOL_DEPTH 10
+#define UBUF_POOL_DEPTH 10
+#define UPUMP_POOL 10
+#define UPUMP_BLOCKER_POOL 10
+#define READ_SIZE 4096
+
+
+/* structure */
+struct uprobe_obe_log {
```
Comment (JDarnley) [inline examples/srt_rx.c:76]: Do we want "obe" things in an upstream program?

File: lib/upipe-srt/upipe_srt_receiver.c [hunk 13]
```diff
@@ -0,0 +1,1454 @@
...
+
+    upipe_srt_receiver_init_upump_mgr(upipe);
+    upipe_srt_receiver_init_upump_timer(upipe);
+    upipe_srt_receiver_init_upump_timer_lost(upipe);
+    upipe_srt_receiver_init_uclock(upipe);
+    upipe_srt_receiver_require_uclock(upipe);
+
+    // FIXME
+    upipe_srt_receiver->socket_id = 0;
+    upipe_srt_receiver->control = NULL;
```
Comment (JDarnley) [inline lib/upipe-srt/upipe_srt_receiver.c:753]: Clarify what the FIXME is about.  I think I saw a comment in handshake that the socket id would be set after receiving the first packet.  If the same applies here maybe you should copy that comment.

File: lib/upipe-srt/upipe_srt_receiver.c [hunk 14]
```diff
@@ -0,0 +1,1454 @@
...
+    total_size -= SRT_HEADER_SIZE;
+
+    static const uint64_t wrap = UINT64_C(4294967296);
+    uint64_t delta = (wrap + ts - (upipe_srt_receiver->previous_ts % wrap)) % wrap;
+    int32_t d32 = delta;
+    bool discontinuity = upipe_srt_receiver->previous_ts == 0;
+
+    uint64_t latency_us = (upipe_srt_receiver->latency * 1000000) / UCLOCK_FREQ;
+    /* Note: d32 is converted to unsigned implictly */
+    if (!discontinuity && (d32 <= latency_us || -d32 <= latency_us)) {
```
Comment (JDarnley) [inline lib/upipe-srt/upipe_srt_receiver.c:1231]: This is still here.  Is this supposed to be an absolute compare?

File: lib/upipe-srt/upipe_srt_receiver.c [hunk 15]
```diff
@@ -0,0 +1,1454 @@
...
+    }
+
+    upipe_verbose_va(upipe, "Data seq %u (retx %u)", seqnum, retransmit);
+    if (d32 < 0 && !discontinuity)
+        uref_clock_set_cr_prog(uref, (upipe_srt_receiver->previous_ts + d32) * UCLOCK_FREQ / 1000000);
+    else
+        uref_clock_set_cr_prog(uref, (upipe_srt_receiver->previous_ts) * UCLOCK_FREQ / 1000000);
+
+    (void)order;
+    (void)num;
```
Comment (JDarnley) [inline lib/upipe-srt/upipe_srt_receiver.c:1251]: Is this just to silence unused variable warnings?  I think you should just comment out the declarations instead.

File: lib/upipe-srt/upipe_srt_sender.c [hunk 16]
```diff
@@ -0,0 +1,952 @@
...
+        upipe_srt_sender_output(upipe_super, uref_dup(uref), NULL);
+
+        diff++;
+        seq += diff;
+        pkts -= diff;
+        if (pkts == 0)
+            return;
+    }
+
+    /* XXX: Is it needed? */
```
Comment (JDarnley) [inline lib/upipe-srt/upipe_srt_sender.c:265]: What in particular?

File: lib/upipe-srt/upipe_srt_sender.c [hunk 17]
```diff
@@ -0,0 +1,952 @@
...
+static void upipe_srt_sender_timer(struct upump *upump)
+{
+    struct upipe *upipe = upump_get_opaque(upump, struct upipe *);
+    struct upipe_srt_sender *upipe_srt_sender = upipe_srt_sender_from_upipe(upipe);
+
+    uint64_t now = uclock_now(upipe_srt_sender->uclock);
+
+    if (now - upipe_srt_sender->last_sent > UCLOCK_FREQ) {
+        struct uref *uref = uref_block_alloc(upipe_srt_sender->uref_mgr,
+                upipe_srt_sender->ubuf_mgr, SRT_HEADER_SIZE + 4 /* WTF */);
```
Comment (JDarnley) [inline lib/upipe-srt/upipe_srt_sender.c:359]: Similar to last time.  What was unexpected?

File: lib/upipe-srt/upipe_srt_sender.c [hunk 18]
```diff
@@ -0,0 +1,952 @@
...
+    srt_set_data_packet_message_number(buf, seqnum);
+    srt_set_data_packet_seq(buf, seqnum);
+    srt_set_data_packet_position(buf, SRT_DATA_POSITION_ONLY);
+    srt_set_data_packet_order(buf, false);
+    srt_set_data_packet_retransmit(buf, false);
+
+#ifdef UPIPE_HAVE_GCRYPT_H
+    if (++upipe_srt_sender->packets_since_key == SRT_KM_PRE_ANNOUNCEMENT_PERIOD) {
+        /* invert the boolean to get the right index */
+        int even_key = !upipe_srt_sender->even_key;
```
Comment (JDarnley) [inline lib/upipe-srt/upipe_srt_sender.c:677]: Would this be better named as "which_key"?  Or just "key_index"?

File: lib/upipe-srt/upipe_srt_sender.c [hunk 19]
```diff
@@ -0,0 +1,952 @@
...
+            upipe_dbg(upipe, "Encryption disabled");
+        }
+
+        upipe_srt_sender->even_key = even_key;
+    } else if (upipe_srt_sender->packets_since_key == 2 * SRT_KM_PRE_ANNOUNCEMENT_PERIOD) {
+        memset(upipe_srt_sender->sek[upipe_srt_sender->even_key], 0, sizeof(upipe_srt_sender->sek[0]));
+    }
+
+    /* invert the boolean to get the right index */
+    int key = !upipe_srt_sender->even_key;
```
Comment (JDarnley) [inline lib/upipe-srt/upipe_srt_sender.c:697]: Same as above.

File: lib/upipe-srt/upipe_srt_sender.c [hunk 20]
```diff
@@ -0,0 +1,952 @@
...
+                ubuf_block_unmap(insert, 0);
+                ubuf_free(insert);
+                uref_free(uref);
+                if (tag)
+                    ubuf_free(tag);
+                return;
+            }
+        }
+
+        //
```
Comment (JDarnley) [inline lib/upipe-srt/upipe_srt_sender.c:803]: Did you have something to add to this empty comment?

**Review by JDarnley — COMMENTED** [review]
Mostly minor nits

File: lib/upipe-srt/upipe_srt_handshake.c [hunk 21]
```diff
@@ -0,0 +1,1773 @@
...
+    struct sockaddr_storage addr;
+    uint64_t establish_time;
+
+    bool expect_conclusion;
+    struct uref *caller_conclusion;
+
+    bool listener;
+
+    uint8_t *stream_id;
+    size_t stream_id_len;
```
[the remark this answers is not in the document]
Comment (funman) [reply lib/upipe-srt/upipe_srt_handshake.c:130]: I am not sure if it matters

**Comment by kylophone** [comment]
What is the status of this SRT implementation? Intended to be mainlined at some point? Experimental?

**Comment by kierank** [comment]
Hi Kyle,

I have been using this implementation in production for nearly a year now
without major issues and interoperating with various libsrt library
versions.

Most of the recent work has been around a few edge cases for rogue callers.
The intention is to mainline it after some review. It's worth saying it's
hard to unit test and also fuzz a complex protocol like SRT so that's a
missing piece of work.

Regards,
Kieran Kunhya

**Comment by kylophone** [comment]
Sounds good. Thanks for the info, Kieran.
